### PR TITLE
updated ExtnasionTime calc

### PR DESCRIPTION
**Pull request by gogogadgetbioprocess**

### CODE REVIEW

File: antha/AnthaStandardLibrary/Packages/enzymes/Polymerases.go [hunk 1]
```diff
@@ -66,6 +68,25 @@ var (
 	}
 )
 
+func CalculateExtensionTime(polymerase *wtype.LHComponent, targetsequence wtype.DNASequence) (wunit.Time, error) {
+
+	var err error = nil
```
Comment (ddn0) [inline antha/AnthaStandardLibrary/Packages/enzymes/Polymerases.go]: var err error

File: antha/AnthaStandardLibrary/Packages/enzymes/Polymerases.go [hunk 2]
```diff
@@ -66,6 +68,25 @@ var (
...
+
+	polymerasename := polymerase.CName
+
+	polymeraseproperties, polymerasefound := DNApolymeraseProperties[polymerasename]
+
+	if !polymerasefound {
+		validoptions := make([]string, 0)
```
Comment (ddn0) [inline antha/AnthaStandardLibrary/Packages/enzymes/Polymerases.go]: var validOptions []string

File: antha/AnthaStandardLibrary/Packages/enzymes/Polymerases.go [hunk 3]
```diff
@@ -24,15 +24,17 @@
...
+	"fmt"
 	"github.com/antha-lang/antha/antha/anthalib/wtype"
 	"github.com/antha-lang/antha/antha/anthalib/wunit"
+	"strings"
 )
 
 var (
 	//
 
 	DNApolymeraseProperties = map[string]map[string]float64{
```
Comment (ddn0) [inline antha/AnthaStandardLibrary/Packages/enzymes/Polymerases.go]: Should these be public?

File: antha/AnthaStandardLibrary/Packages/enzymes/Polymerases.go [hunk 4]
```diff
@@ -66,6 +68,25 @@ var (
...
+
+	polymeraseproperties, polymerasefound := DNApolymeraseProperties[polymerasename]
+
+	if !polymerasefound {
+		validoptions := make([]string, 0)
+		for key, _ := range DNApolymeraseProperties {
+			validoptions = append(validoptions, key)
+		}
+
+		err = fmt.Errorf("No Properties for " + polymerasename + " found." + " Valid options are: " + strings.Join(validoptions, ","))
```
Comment (ddn0) [inline antha/AnthaStandardLibrary/Packages/enzymes/Polymerases.go]: return here

also

refactor to error type:
```go
type invalidProperty struct { }

var anInvalidProperty = &invalidProperty{}

func (ip *invalidProperty) Error() string {
  var props []string
  for prop := range DNAPolymeraseProperties {
    props = append(props, key)
  }
  sort.Strings(props)
  return "Valid options are: " + strings.Join(props, ",")
} 
...

func CalculateExtensionTime
  ...
  return wunit.Time{}, anInvalidProperty
  ...
```

File: antha/AnthaStandardLibrary/Packages/enzymes/Polymerases.go [hunk 5]
```diff
@@ -66,6 +68,25 @@ var (
 	}
 )
 
+func CalculateExtensionTime(polymerase *wtype.LHComponent, targetsequence wtype.DNASequence) (wunit.Time, error) {
```
Comment (ddn0) [inline antha/AnthaStandardLibrary/Packages/enzymes/Polymerases.go]: targetSequence

**Comment by ddn0** [comment]
@synthacebot test this please

**Comment by gogogadgetbioprocess** [comment]
all changes done (including making the map private)

File: antha/AnthaStandardLibrary/Packages/enzymes/Polymerases.go [hunk 6]
```diff
@@ -66,6 +81,19 @@ var (
...
+
+	polymerasename := polymerase.CName
+
+	polymeraseproperties, polymerasefound := dnaPolymeraseProperties[polymerasename]
+
+	if !polymerasefound {
+
+		return wunit.Time{}, anInvalidProperty
+	}
+	return wunit.NewTime(float64(len(targetSequence.Sequence()))/polymeraseproperties["SperKb_lower"], "s"), anInvalidProperty
```
Comment (ddn0) [inline antha/AnthaStandardLibrary/Packages/enzymes/Polymerases.go]: Checking if SperKb_lower exists in polymeraseProperties is preferable to divide by zero errors.

**Comment by gogogadgetbioprocess** [comment]
@synthacebot test this please

**Comment by ddn0** [comment]
@synthacebot test this please

**Comment by gogogadgetbioprocess** [comment]
@synthacebot test this please

**Comment by gogogadgetbioprocess** [comment]
@synthacebot test this please

**Pull request merged**